### ISSUE TITLE
[BUG] 3006.7 broke my setup. Doesn't follow symlinks

### ISSUE THREAD

**Issue by stefancastille** [saltstack/salt#66133]
**Description**
After an update to 3006.7 the salt-master could no longer apply states to minions. The following error was returned:

     **Comment: No Top file or master_tops data matches found. Please see master log for details.**

Minions could still be contacted and commands worked (e.g. salt minion cmd.run hostname)

Logs did not show any information at all indicating what the problem could be, not even when running loglevel trace.

**Setup**
(Please provide relevant configs and/or SLS files (be sure to remove sensitive info. There is no general set-up of Salt.)

Please be as specific as possible and give set-up details.

- [X] on-prem machine
- [X] VM ESXi
- [ ] VM running on a cloud service, please be explicit and add details
- [ ] container (Kubernetes, Docker, containerd, etc. please specify)
- [ ] or a combination, please be explicit
- [ ] jails if it is FreeBSD
- [X] classic packaging - apt install on ubuntu 22.04
- [ ] onedir packaging
- [ ] used bootstrap to install

**Steps to Reproduce the behavior**
Have the salt state config in /opt/config, make a symlink at /src/salt/top.sls pointing to /opt/config/top.sls
run **salt <minion> state.apply** 

**Expected behavior**
The state is applied correctly.

**Screenshots**

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.)

```yaml
Salt Version:
          Salt: 3006.7

Python Version:
        Python: 3.10.13 (main, Feb 19 2024, 03:31:20) [GCC 11.2.0]

Dependency Versions:
          cffi: 1.14.6
      cherrypy: unknown
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.1.3
       libgit2: Not Installed
  looseversion: 1.0.2
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     packaging: 22.0
     pycparser: 2.21
      pycrypto: Not Installed
  pycryptodome: 3.19.1
        pygit2: Not Installed
  python-gnupg: 0.4.8
        PyYAML: 6.0.1
         PyZMQ: 23.2.0
        relenv: 0.15.1
         smmap: Not Installed
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: 4.3.4

System Versions:
          dist: ubuntu 22.04.4 jammy
        locale: utf-8
       machine: x86_64
       release: 5.15.0-97-generic
        system: Linux
       version: Ubuntu 22.04.4 jammy
```
</details>

**Additional context**
If i copy top.sls to the location the error goes to the next level, it seems like it does not follow any symlinks at all. This was not the behaviour in the previous version.
```
minion:
    Data failed to compile:
----------
    No matching sls found for 'generic-config-internal' in env 'base'

```
Removing the symlinks and simply going direct files breaks other things so not really an easy fix.

**Comment by hurzhurz** [comment]
Got the same issue. Pretty sure it is caused by this change in roots.py from 3006.6:  [e0cdb80](https://github.com/saltstack/salt/commit/e0cdb80b55123f4a024759ffcf2b3f0e0788e7ab#diff-1f96c1bf1efb4edbad6f6ea24768d127521489e238b3e5ff466a1c56323257d6)

I have not checked this in depth, but probably `salt.utils.verify.clean_path()` doesn't like symlink, respectively resolves the real path and compares it against the roots directory, what fails.

Quick workaround would be to take the roots.py from 3006.5 (or older) and place it in `/srv/salt/_fileserver` (+ "salt-run saltutil.sync_fileserver" & salt-master restart):
**https://raw.githubusercontent.com/saltstack/salt/v3006.5/salt/fileserver/roots.py**

**Comment by dwoz** [comment]
> Got the same issue. Pretty sure it is caused by this change in roots.py from 3006.6: [e0cdb80](https://github.com/saltstack/salt/commit/e0cdb80b55123f4a024759ffcf2b3f0e0788e7ab#diff-1f96c1bf1efb4edbad6f6ea24768d127521489e238b3e5ff466a1c56323257d6)
> 
> I have not checked this in depth, but probably `salt.utils.verify.clean_path()` doesn't like symlink, respectively resolves the real path and compares it against the roots directory, what fails.
> 
> Quick workaround would be to take the roots.py from 3006.5 (or older) and place it in `/srv/salt/_fileserver` (+ "salt-run saltutil.sync_fileserver" & salt-master restart): **https://raw.githubusercontent.com/saltstack/salt/v3006.5/salt/fileserver/roots.py**

That's what I was guessing too. The changes made in 3006.6 were intentional. However, I do understand the reasoning behind wanting symlinks to work.

**Comment by pvarlien** [comment]
Appears to be broken in 3007.0 as well.

**Comment by alan-cugler** [comment]
Just tried using symlinked states directory defined with file_roots in the salt master. ignores the symlink in its entirety still. I am using 3007.

**Comment by rjel2159** [comment]
> > Got the same issue. Pretty sure it is caused by this change in roots.py from 3006.6: [e0cdb80](https://github.com/saltstack/salt/commit/e0cdb80b55123f4a024759ffcf2b3f0e0788e7ab#diff-1f96c1bf1efb4edbad6f6ea24768d127521489e238b3e5ff466a1c56323257d6)
> > I have not checked this in depth, but probably `salt.utils.verify.clean_path()` doesn't like symlink, respectively resolves the real path and compares it against the roots directory, what fails.
> > Quick workaround would be to take the roots.py from 3006.5 (or older) and place it in `/srv/salt/_fileserver` (+ "salt-run saltutil.sync_fileserver" & salt-master restart): **https://raw.githubusercontent.com/saltstack/salt/v3006.5/salt/fileserver/roots.py**
> 
> That's what I was guessing too. The changes made in 3006.6 were intentional. However, I do understand the reasoning behind wanting symlinks to work.

Can someone please point me to this change mentioned in the 3006.6 release notes, the commit & relevant issue number opened for it to justify the need to do this?

**Comment by rjel2159** [comment]
> > > Got the same issue. Pretty sure it is caused by this change in roots.py from 3006.6: [e0cdb80](https://github.com/saltstack/salt/commit/e0cdb80b55123f4a024759ffcf2b3f0e0788e7ab#diff-1f96c1bf1efb4edbad6f6ea24768d127521489e238b3e5ff466a1c56323257d6)
> > > I have not checked this in depth, but probably `salt.utils.verify.clean_path()` doesn't like symlink, respectively resolves the real path and compares it against the roots directory, what fails.
> > > Quick workaround would be to take the roots.py from 3006.5 (or older) and place it in `/srv/salt/_fileserver` (+ "salt-run saltutil.sync_fileserver" & salt-master restart): **https://raw.githubusercontent.com/saltstack/salt/v3006.5/salt/fileserver/roots.py**
> > 
> > 
> > That's what I was guessing too. The changes made in 3006.6 were intentional. However, I do understand the reasoning behind wanting symlinks to work.
> 
> We've just run into this exact issue ourselves when trying to upgrade from 3005.1 to 3007.0. Can someone please point me to this change mentioned in the 3006.6 release notes & relevant issue number opened for it to justify the need to do this?

Btw, I can confirm the following w/a fixes our issue:

```
# diff -u lib/python3.9/site-packages/salt/utils/verify.py{,.new}
--- lib/python3.9/site-packages/salt/utils/verify.py    2024-03-26 15:35:40.794310801 -0400
+++ lib/python3.9/site-packages/salt/utils/verify.py.new        2024-03-26 15:35:28.171586128 -0400
@@ -540,7 +540,8 @@
     else:
         if os.path.dirname(real_path) == os.path.normpath(real_root):
             return real_path
-    return ""
+    #return ""
+    return(real_path)


 def valid_id(opts, id_):
```

**Comment by PaulChristophel** [comment]
Another option is to update the function in question to allow for "safe" user defined paths:
```python
    allowed_paths = ["/salt-formula"]
    for allowed_path in allowed_paths:
        real_allowed_path = _realpath(allowed_path)
        if real_path.startswith(real_allowed_path):
            return real_path
    return ""
```

**Comment by vshalchian** [comment]
I do use symlinks a lot in my salt directory and since bug , they are all broken and it is preventing me from upgrading to newer versions. I might end up using roots.py from previous versions in the meantime to unblock the version upgrade, but would be nice to have this one fixed

**Comment by alan-cugler** [comment]
## I would argue more power to the end user. 

If there was a good reason to not allow symlinks in this way, then lets document that and then make `allow_symlink_in_fileroots = True` in the `master.conf` settings. Now its a feature that can be enabled and they can document the main downside of why they removed that ability by default like we do for all other features that have a reasonable downside for utilizing.

**Comment by hurzhurz** [comment]
@alan-cugler I'm pretty sure it was not intended to disallow symlinks, it is just a side effect of the fix for critical CVE: https://saltproject.io/security-announcements/2024-01-31-advisory/index.html

And by the way, there is already a similar setting [fileserver_followsymlinks](https://docs.saltproject.io/en/latest/ref/configuration/master.html#fileserver-followsymlinks). While it has another purpose, it could be useful for optional weakening of the path check.

---
Here is a suggestion for a slightly relaxed version of the [clean_path](https://github.com/saltstack/salt/blob/master/salt/utils/verify.py#L527-L546) function, which is used to verify the path:

```python
def clean_path(root, path, subdir=False):
    if not os.path.isabs(root):
        return ""
    root = os.path.normpath(root)
    if not os.path.isabs(path):
        path = os.path.join(root, path)
    path = os.path.normpath(path)
    if subdir:
        if os.path.commonpath([path, root]) == root:
            return path
    else:
        if os.path.dirname(path) == root:
            return path
    return ""
```

From my testing, it seems to fix the CVE while symlinks still work.

**Comment by dariosusman** [comment]
This has impacted on 3005.5 badly as well. We do use to follow symlinks for let'sencrypt/certbot key/cert pairs deployments.
That bit of code worked very well. Thank you!

**Comment by rjel2159** [comment]
> @alan-cugler I'm pretty sure it was not intended to disallow symlinks, it is just a side effect of the fix for a critical CVE: https://saltproject.io/security-announcements/2024-01-31-advisory/index.html
> 
> And by the way, there is already a similar setting [fileserver_followsymlinks](https://docs.saltproject.io/en/latest/ref/configuration/master.html#fileserver-followsymlinks). While it has another purpose, it could be useful for optional weakening of the path check.
> 
> Here is a suggestion for a slightly relaxed version of the [clean_path](https://github.com/saltstack/salt/blob/master/salt/utils/verify.py#L527-L546) function, which is used to verify the path:
> 
> ```python
> def clean_path(root, path, subdir=False):
>     if not os.path.isabs(root):
>         return ""
>     root = os.path.normpath(root)
>     if not os.path.isabs(path):
>         path = os.path.join(root, path)
>     path = os.path.normpath(path)
>     if subdir:
>         if os.path.commonpath([path, root]) == root:
>             return path
>     else:
>         if os.path.dirname(path) == root:
>             return path
>     return ""
> ```
> 
> From my testing, it seems to fix the CVE while symlinks still work.

Thanks, this proposed fix works for us as well.

**Comment by jtlarson** [comment]
> @alan-cugler I'm pretty sure it was not intended to disallow symlinks, it is just a side effect of the fix for a critical CVE: https://saltproject.io/security-announcements/2024-01-31-advisory/index.html
> 
> And by the way, there is already a similar setting [fileserver_followsymlinks](https://docs.saltproject.io/en/latest/ref/configuration/master.html#fileserver-followsymlinks). While it has another purpose, it could be useful for optional weakening of the path check.
> 
> Here is a suggestion for a slightly relaxed version of the [clean_path](https://github.com/saltstack/salt/blob/master/salt/utils/verify.py#L527-L546) function, which is used to verify the path:
> 
> ```python
> def clean_path(root, path, subdir=False):
>     if not os.path.isabs(root):
>         return ""
>     root = os.path.normpath(root)
>     if not os.path.isabs(path):
>         path = os.path.join(root, path)
>     path = os.path.normpath(path)
>     if subdir:
>         if os.path.commonpath([path, root]) == root:
>             return path
>     else:
>         if os.path.dirname(path) == root:
>             return path
>     return ""
> ```
> 
> From my testing, it seems to fix the CVE while symlinks still work.

@hurzhurz I'm trying to work around this same issue on 3005.5, but salt is still complaining about the symlink after replacing the clean_path function with your suggested code. Do you have any suggestions on how I might need to modify your code to work with [the 3005.5 verify.py](https://github.com/saltstack/salt/blob/e39b65aa4be1675b9bb156aba1c2deeababa0420/salt/utils/verify.py#L512-L531), or do I need to modify other files beyond the verify.py in order to re-enable symlinks? 

More generally--can anyone confirm if symlinks are going to be formally re-enabled in the future? And what are the best alternatives for including files that we don't want to be located in the salt base?

**Comment by vshalchian** [comment]
@jtlarson Looks like there is a PR to fix #66133
https://github.com/saltstack/salt/pull/66419
It is merged, though I could not found it on any of the recent releases , I might be wrong about that though

**Comment by hurzhurz** [comment]
@jtlarson I have not tried to modify the verify.py as a workaround. I'm not sure if that is easily possible. It might depend on the installation method or package type (classic / onedir).
A simpler workaround, which I actually used myself, is what I mentioned here: https://github.com/saltstack/salt/issues/66133#issuecomment-1964013273

As @vshalchian said, I made a PR for a fix and it is already merged.
For 3007 it is included in the latest release (3007.1).
There was no new 3006 release since. I guess it will be included in 3006.9 (if that won't be a CVE only release).
I don't think it will ever be in a 3005 release. As I understand, 3005 does only receive CVE fixes and nothing else.
